torchmultimodal/modules/layers/position_embedding.py

```diff
@@ -169,3 +169,105 @@ def forward(self, t: Tensor) -> Tensor:
         if self.embed_dim % 2 == 1:
             embeddings = nn.functional.pad(embeddings, (0, 1))
         return embeddings
+
+
+class RotaryPositionalEmbeddings(nn.Module):
+    def __init__(
+        self,
+        dim: int,
+        max_position_embeddings: Union[int, float] = 2048,
+        ratio: int = 10000,
+        device: torch.device = None,
+    ):
+        """
+        Implements Rotary Positional Embeddings (RoPE)
+        proposed in: https://arxiv.org/abs/2104.09864
+
+        Args
+        ----
+        dim : int
+            Per-head embedding dimension
+        max_position_embeddings : int
+            Maximum expected sequence length for the model, if exceeded the cached freqs will be recomputed
+        ratio: int
+            The ratio for the geometric progression to compute the rotation angles
+        """
+        super().__init__()
+        self.register_buffer(
+            "freqs",
+            1.0
+            / (
+                ratio
+                ** (torch.arange(0, dim, 2, device=device)[: (dim // 2)].float() / dim)
+            ),
+        )
+        self.compute_freqs_cis(max_position_embeddings)
+
+    def compute_freqs_cis(
```

Comment:
Random q: what does cis mean here?

Reply:
it's short form for rotation transform technically doing e^(alpha*i) = cos(alpha) + i * sin(alpha), or shortened, **c**os + **i** * **s**in = cis.

Reply:
should probably add that in the docstring actually, otherwise it's too cryptic.